File: ecosystem/python/sdk/aptos_sdk/account_sequence_number.py
# Copyright © Aptos Foundation
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import asyncio
import logging
from typing import Callable, Optional

from aptos_sdk.account_address import AccountAddress
from aptos_sdk.async_client import ApiError, RestClient


class AccountSequenceNumberConfig:
    """Common configuration for account number generation"""

    maximum_in_flight: int = 100
    maximum_wait_time: int = 30
    sleep_time: float = 0.01


class AccountSequenceNumber:
    """
    A managed wrapper around sequence numbers that implements the trivial flow control used by the
    Aptos faucet:
    * Submit up to 100 transactions per account in parallel with a timeout of 20 seconds
    * If local assumes 100 are in flight, determine the actual committed state from the network
    * If there are less than 100 due to some being committed, adjust the window
    * If 100 are in flight Wait .1 seconds before re-evaluating
    * If ever waiting more than 30 seconds restart the sequence number to the current on-chain state
    Assumptions:
    * Accounts are expected to be managed by a single AccountSequenceNumber and not used otherwise.
    * They are initialized to the current on-chain state, so if there are already transactions in
      flight, they may take some time to reset.
    * Accounts are automatically initialized if not explicitly

    Notes:
    * This is co-routine safe, that is many async tasks can be reading from this concurrently.
    * The state of an account cannot be used across multiple AccountSequenceNumber services.
    * The synchronize method will create a barrier that prevents additional next_sequence_number
      calls until it is complete.
    * This only manages the distribution of sequence numbers it does not help handle transaction
      failures.
    * If a transaction fails, you should call synchronize and wait for timeouts.
    * Mempool limits the number of transactions per account to 100, hence why we chose 100.
    """

    _client: RestClient
    _account: AccountAddress
    _lock: asyncio.Lock

    _maximum_in_flight: int = 100
    _maximum_wait_time: int = 30
    _sleep_time: float = 0.01

    _last_committed_number: Optional[int]
    _current_number: Optional[int]

    def __init__(
        self,
        client: RestClient,
        account: AccountAddress,
        config: AccountSequenceNumberConfig = AccountSequenceNumberConfig(),
    ):
        self._client = client
        self._account = account
        self._lock = asyncio.Lock()

        self._last_uncommitted_number = None
        self._current_number = None

        self._maximum_in_flight = config.maximum_in_flight
        self._maximum_wait_time = config.maximum_wait_time
        self._sleep_time = config.sleep_time

    async def next_sequence_number(self, block: bool = True) -> Optional[int]:
        """
        Returns the next sequence number available on this account. This leverages a lock to
        guarantee first-in, first-out ordering of requests.
        """
        async with self._lock:
            if self._last_uncommitted_number is None or self._current_number is None:
                await self._initialize()
            # If there are more than self._maximum_in_flight in flight, wait for a slot.
            # Or at least check to see if there is a slot and exit if in non-blocking mode.
            if (
                self._current_number - self._last_uncommitted_number
                >= self._maximum_in_flight
            ):
                await self._update()
                if (
                    self._current_number - self._last_uncommitted_number
                    >= self._maximum_in_flight
                ):
                    if not block:
                        return None
                    await self._resync(
                        lambda acn: acn._current_number - acn._last_uncommitted_number
                        >= acn._maximum_in_flight
                    )

            next_number = self._current_number
            self._current_number += 1
        return next_number

    async def _initialize(self):
        """Optional initializer. called by next_sequence_number if not called prior."""
        self._current_number = await self._current_sequence_number()
        self._last_uncommitted_number = self._current_number

    async def synchronize(self):
        """
        Poll the network until all submitted transactions have either been committed or until
        the maximum wait time has elapsed. This will prevent any calls to next_sequence_number
        until this called has returned.
        """
        async with self._lock:
            await self._update()
            await self._resync(
                lambda acn: acn._last_uncommitted_number != acn._current_number
            )

    async def _resync(self, check: Callable[[AccountSequenceNumber], bool]):
        """Forces a resync with the upstream, this should be called within the lock"""
        start_time = await self._client.current_timestamp()
        failed = False
        while check(self):
            ledger_time = await self._client.current_timestamp()
            if ledger_time - start_time > self._maximum_wait_time:
                logging.warn(
                    f"Waited over {self._maximum_wait_time} seconds for a transaction to commit, resyncing {self._account}"
                )
                failed = True
                break
            else:
                await asyncio.sleep(self._sleep_time)
                await self._update()
        if not failed:
            return
        for seq_num in range(self._last_uncommitted_number + 1, self._current_number):
            while True:
                try:
                    result = (
                        await self._client.account_transaction_sequence_number_status(
                            self._account, seq_num
                        )
                    )
                    if result:
                        break
                except ApiError as error:
                    if error.status_code == 404:
                        break
                    raise
        await self._initialize()

    async def _update(self):
        self._last_uncommitted_number = await self._current_sequence_number()
        return self._last_uncommitted_number

    async def _current_sequence_number(self) -> int:
        return await self._client.account_sequence_number(self._account)


import unittest
import unittest.mock


class Test(unittest.IsolatedAsyncioTestCase):
    async def test_common_path(self):
        """
        Verifies that:
        * AccountSequenceNumber returns sequential numbers starting from 0
        * When the account has been updated on-chain include that in computations 100 -> 105
        * Ensure that none is returned if the call for next_sequence_number would block
        * Ensure that synchronize completes if the value matches on-chain
        """
        patcher = unittest.mock.patch(
            "aptos_sdk.async_client.RestClient.account_sequence_number", return_value=0
        )
        patcher.start()

        rest_client = RestClient("https://fullnode.devnet.aptoslabs.com/v1")
        account_sequence_number = AccountSequenceNumber(
            rest_client, AccountAddress.from_hex("b0b")
        )
        last_seq_num = 0
        for seq_num in range(5):
            last_seq_num = await account_sequence_number.next_sequence_number()
            self.assertEqual(last_seq_num, seq_num)

        patcher.stop()
        patcher = unittest.mock.patch(
            "aptos_sdk.async_client.RestClient.account_sequence_number", return_value=5
        )
        patcher.start()

        for seq_num in range(AccountSequenceNumber._maximum_in_flight):
            last_seq_num = await account_sequence_number.next_sequence_number()
            self.assertEqual(last_seq_num, seq_num + 5)

        self.assertEqual(
            await account_sequence_number.next_sequence_number(block=False), None
        )
        next_sequence_number = last_seq_num + 1
        patcher.stop()
        patcher = unittest.mock.patch(
            "aptos_sdk.async_client.RestClient.account_sequence_number",
            return_value=next_sequence_number,
        )
        patcher.start()

        self.assertNotEqual(account_sequence_number._current_number, last_seq_num)
        await account_sequence_number.synchronize()
        self.assertEqual(account_sequence_number._current_number, next_sequence_number)

File: ecosystem/typescript/sdk/src/transactions/account_sequence_number.ts
/**
 * A wrapper that handles and manages an account sequence number.
 *
 * Submit up to `maximumInFlight` transactions per account in parallel with a timeout of `sleepTime`
 * If local assumes `maximumInFlight` are in flight, determine the actual committed state from the network
 * If there are less than `maximumInFlight` due to some being committed, adjust the window
 * If `maximumInFlight` are in flight, wait `sleepTime` seconds before re-evaluating
 * If ever waiting more than `maxWaitTime` restart the sequence number to the current on-chain state
 *
 * Assumptions:
 * Accounts are expected to be managed by a single AccountSequenceNumber and not used otherwise.
 * They are initialized to the current on-chain state, so if there are already transactions in
 * flight, they may take some time to reset.
 * Accounts are automatically initialized if not explicitly
 *
 * Notes:
 * This is co-routine safe, that is many async tasks can be reading from this concurrently.
 * The state of an account cannot be used across multiple AccountSequenceNumber services.
 * The synchronize method will create a barrier that prevents additional nextSequenceNumber
 * calls until it is complete.
 * This only manages the distribution of sequence numbers it does not help handle transaction
 * failures.
 * If a transaction fails, you should call synchronize and wait for timeouts.
 */

import { AptosAccount } from "../account";
import { Provider } from "../providers";
import { sleep } from "../utils";

// returns `now` time in seconds
const now = () => Math.floor(Date.now() / 1000);

export class AccountSequenceNumber {
  readonly provider: Provider;

  readonly account: AptosAccount;

  // sequence number on chain
  lastUncommintedNumber: bigint | null = null;

  // local sequence number
  currentNumber: bigint | null = null;

  /**
   * We want to guarantee that we preserve ordering of workers to requests.
   *
   * `lock` is used to try to prevent multiple coroutines from accessing a shared resource at the same time,
   * which can result in race conditions and data inconsistency.
   * This code actually doesn't do it though, since we aren't giving out a slot, it is still somewhat a race condition.
   *
   * The ideal solution is likely that each thread grabs the next number from a incremental integer.
   * When they complete, they increment that number and that entity is able to enter the `lock`.
   * That would guarantee ordering.
   */
  lock = false;

  maxWaitTime: number;

  maximumInFlight: number;

  sleepTime: number;

  constructor(
    provider: Provider,
    account: AptosAccount,
    maxWaitTime: number,
    maximumInFlight: number,
    sleepTime: number,
  ) {
    this.provider = provider;
    this.account = account;
    this.maxWaitTime = maxWaitTime;
    this.maximumInFlight = maximumInFlight;
    this.sleepTime = sleepTime;
  }

  /**
   * Returns the next available sequence number for this account
   *
   * @returns next available sequence number
   */
  async nextSequenceNumber(): Promise<bigint | null> {
    /* eslint-disable no-await-in-loop */
    while (this.lock) {
      await sleep(this.sleepTime);
    }

    this.lock = true;
    let nextNumber = BigInt(0);
    try {
      if (this.lastUncommintedNumber === null || this.currentNumber === null) {
        await this.initialize();
      }

      if (this.currentNumber! - this.lastUncommintedNumber! >= this.maximumInFlight) {
        await this.update();

        const startTime = now();
        while (this.currentNumber! - this.lastUncommintedNumber! >= this.maximumInFlight) {
          await sleep(this.sleepTime);
          if (now() - startTime > this.maxWaitTime) {
            /* eslint-disable no-console */
            console.warn(`Waited over 30 seconds for a transaction to commit, resyncing ${this.account.address()}`);
            await this.initialize();
          } else {
            await this.update();
          }
        }
      }
      nextNumber = this.currentNumber!;
      this.currentNumber! += BigInt(1);
    } catch (e) {
      console.error("error in getting next sequence number for this account", e);
    } finally {
      this.lock = false;
    }
    return nextNumber;
  }

  /**
   * Initializes this account with the sequence number on chain
   */
  async initialize(): Promise<void> {
    const { sequence_number: sequenceNumber } = await this.provider.getAccount(this.account.address());
    this.currentNumber = BigInt(sequenceNumber);
    this.lastUncommintedNumber = BigInt(sequenceNumber);
  }

  /**
   * Updates this account sequence number with the one on-chain
   *
   * @returns on-chain sequence number for this account
   */
  async update(): Promise<bigint> {
    const { sequence_number: sequenceNumber } = await this.provider.getAccount(this.account.address());
    this.lastUncommintedNumber = BigInt(sequenceNumber);
    return this.lastUncommintedNumber;
  }

  /**
   * Synchronizes local sequence number with the seqeunce number on chain for this account.
   *
   * Poll the network until all submitted transactions have either been committed or until
   * the maximum wait time has elapsed
   */
  async synchronize(): Promise<void> {
    if (this.lastUncommintedNumber === this.currentNumber) return;

    /* eslint-disable no-await-in-loop */
    while (this.lock) {
      await sleep(this.sleepTime);
    }

    this.lock = true;

    try {
      await this.update();
      const startTime = now();
      while (this.lastUncommintedNumber !== this.currentNumber) {
        if (now() - startTime > this.maxWaitTime) {
          /* eslint-disable no-console */
          console.warn(`Waited over 30 seconds for a transaction to commit, resyncing ${this.account.address()}`);
          await this.initialize();
        } else {
          await sleep(this.sleepTime);
          await this.update();
        }
      }
    } catch (e) {
      console.error("error in synchronizing this account sequence number with the one on chain", e);
    } finally {
      this.lock = false;
    }
  }
}
